newton/_src/sim/builder.py

```
@@ -6369,23 +6437,42 @@ def color(
+                bending_edge_active_mask,
+                tet_indices,
+                tet_materials[:, 0] * tet_materials[:, 1] if len(tet_materials) else None,
             )
-        else:
-            # No edges to color - assign all particles to single color group
-            if len(self.particle_q) > 0:
-                self.particle_color_groups = [np.arange(len(self.particle_q), dtype=int)]
+
+            if len(graph_edge_indices) > 0:
+                self.particle_color_groups = color_graph(
+                    num_nodes,
```

Ruff (0.14.14)

[error] Undefined name `num_nodes`

(F821)